kagi/utils/webauthn.py

```diff
@@ -0,0 +1,180 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import json
+
+import webauthn as pywebauthn
+from webauthn.helpers import base64url_to_bytes, generate_challenge
+from webauthn.helpers.exceptions import (
+    InvalidAuthenticationResponse,
+    InvalidRegistrationResponse,
+)
+from webauthn.helpers.options_to_json import options_to_json
+from webauthn.helpers.structs import (
+    AttestationConveyancePreference,
+    AuthenticationCredential,
+    AuthenticatorSelectionCriteria,
+    AuthenticatorTransport,
+    PublicKeyCredentialDescriptor,
+    RegistrationCredential,
+    UserVerificationRequirement,
+)
+
+
+class AuthenticationRejectedError(Exception):
+    pass
+
+
+class RegistrationRejectedError(Exception):
+    pass
+
+
+def _get_webauthn_user_public_key_credential_descriptors(user, *, rp_id):
+    """
+    Returns a webauthn.WebAuthnUser instance corresponding
+    to the given user model, with properties suitable for
+    usage within the webauthn API.
+    """
+    return [
+        PublicKeyCredentialDescriptor(
+            id=base64url_to_bytes(credential.credential_id),
+            transports=[
+                AuthenticatorTransport.USB,
+                AuthenticatorTransport.NFC,
+                AuthenticatorTransport.BLE,
+                AuthenticatorTransport.INTERNAL,
+            ],
+        )
+        for credential in user.webauthn_keys.all()
+    ]
+
+
+def _get_webauthn_user_public_keys(user, *, rp_id):
+    return [
+        (
+            base64url_to_bytes(credential.public_key),
+            credential.sign_count,
+        )
+        for credential in user.webauthn_keys.all()
+    ]
+
+
+def _webauthn_b64encode(source):
+    return base64.urlsafe_b64encode(source).rstrip(b"=")
+
+
+def generate_webauthn_challenge():
+    """
+    Returns a random challenge suitable for use within
+    Webauthn's credential and configuration option objects.
+
+    See: https://w3c.github.io/webauthn/#cryptographic-challenges
+    """
+    return generate_challenge()
+
+
+def get_credential_options(user, *, challenge, rp_name, rp_id):
+    """
+    Returns a dictionary of options for credential creation
+    on the client side.
+    """
+    _authenticator_selection = AuthenticatorSelectionCriteria()
+    _authenticator_selection.user_verification = UserVerificationRequirement.DISCOURAGED
```

Comment:
If you want to allow the user to register new `usb` `nfc` yubikey. One that its pin code wasn't defined before,
 you should define the 
`_authenticator_selection.resident_key=ResidentKeyRequirement.PREFERRED` [see](https://github.com/duo-labs/webauthn.io/blob/master/_app/homepage/services/registration.py#L48:~:text=resident_key%3DResidentKeyRequirement.PREFERRED)